examples/enc-search-sample.rs

```diff
@@ -0,0 +1,421 @@
+// words, decrypt the name in the customer record, then check whether the
+// decrypted name contains all of the word fragments from the query string.
+// If so, return the decrypted name. Otherwise, the record was a false positive,
+// so return None.
+// start-snippet{filterCust}
+async fn filter_customer(
+    sdk: &IronOxide,
+    cust: &Customer,
+    name_parts: &Vec<&str>,
+) -> Result<Option<String>> {
+    let cust_enc_name = base64::decode(&cust.name)?;
+    let cust_name_keys = base64::decode(&cust.name_keys)?;
+    let dec_result = sdk
+        .document_decrypt_unmanaged(&cust_enc_name, &cust_name_keys)
+        .await?;
+    let dec_name = std::str::from_utf8(&dec_result.decrypted_data()).unwrap();
+    let dec_name_trans = ironoxide::search::transliterate_string(&dec_name);
+    if name_parts
+        .iter()
+        .all(|name_part| dec_name_trans.contains(name_part))
+    {
+        Ok(Some(dec_name.to_string()))
+    } else {
+        Ok(None)
+    }
+}
+// end-snippet{filterCust}
+
+// Given a query string, generate the set of index tokens and use this to
+// retrieve possible matches from the server. Transliterate the string,
+// break it into pieces on white space, and for each returned customer,
+// check to see if the customer name actually contains the words from the
+// query. If so, output the customer ID and name.
+// start-snippet{displayCust}
+async fn display_matching_customers(
+    sdk: &IronOxide,
+    name_index: &BlindIndexSearch,
+    query_str: &str,
+) -> Result<()> {
+    let query_tokens = name_index
+        .tokenize_query(query_str, None)?
+        .into_iter()
+        .collect();
+    let customer_recs = search_customers(&query_tokens);
+    let trans_query = ironoxide::search::transliterate_string(&query_str);
+    let name_parts: Vec<&str> = trans_query.split_whitespace().collect();
+    for cust in customer_recs.iter() {
+        let result = filter_customer(&sdk, &cust, &name_parts).await?;
+        match result {
+            Some(decrypted_name) => println!("{} {} matched query", cust.id, decrypted_name),
+            None => println!("{} did not match query", cust.id),
+        }
+    }
+    Ok(())
+}
+// end-snippet{displayCust}
+
+// Mock out some functions that would call the back-end service and return data in the real app
+// We use some mutable global state to persist things that the service usually would.
+struct GlobalState {
+    pub encrypted_salt: String,
+    pub customers: Vec<Customer>,
+    pub start_time: String,
+}
+
+impl GlobalState {
+    pub fn new() -> Self {
+        GlobalState {
+            encrypted_salt: "".to_string(),
+            customers: vec![],
+            start_time: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs()
+                .to_string(),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLOBAL_STATE: MutStatic<GlobalState> = MutStatic::new();
```

Comment:
If it were up to me, I would just pass this state to the functions where it's needed. The MutStatic adds more noise than it's worth IMO.

Reply:
Figured out a better initialization.